### PR TITLE
OIDC required claims

### LINKED ISSUE #61276: OIDC flag to check hosted_domain in id-token
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

When OIDC is enabled and configured, we can authenticate on the Kubernetes API via Google Auth with any gmail address.

**What you expected to happen**:

We should be able to limit access only to hosted domain addresses.
We could add an --oidc-hosted-domain flag for the API server, with the authorized domain in parameter. The authentication plugin would check the domain against the `hd` field in the id-token payload.

**How to reproduce it (as minimally and precisely as possible)**:

- Create Google OAuth clientSecret and clientID.
- Authenticate via Google and get OIDC token.
- Update kubeconfig with a new user (the google authenticated user).

Any gmail account can authenticate.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): Debian
- Kernel (e.g. `uname -a`): Linux 4.4.111-k8s
- Install tools: kops
- Others:

/sig auth

### PR DESCRIPTION
**Pull request by rithujohn191**
**What this PR does / why we need it**: 
Currently there is no mechanism for a user to specify claims in the OIDC authentication process that are required to be present in the ID Token with an expected value. This PR adds the required claims support for the OIDC authentication. It allows users to pass in a `--oidc-required-claims` flag, and key=value pairs in the API config, which will ensure that the specified `required claims` are checked against the ID Token claims.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61276

**Special notes for your reviewer**:
Ran the following commands to update godep files:

```
./hack/godep-restore.sh -v
./hack/godep-save.sh
./hack/update-staging-godeps.sh
./hack/update-bazel.sh
```
Since we don't officially support go 1.10, kept go version to 1.9

**Release note**:

```release-note
kube-apiserver: oidc authentication now supports requiring specific claims with `--oidc-required-claim=<claim>=<value>`
```
/sig auth
/kind feature
/assign @ericchiang

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/Godeps/Godeps.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
 	"ImportPath": "k8s.io/apimachinery",
-	"GoVersion": "go1.9",
```
Comment (ericchiang) [inline staging/src/k8s.io/apimachinery/Godeps/Godeps.json]: Need to revert this change.

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 2]
```diff
@@ -323,6 +329,25 @@ func (a *Authenticator) AuthenticateToken(token string) (user.Info, bool, error)
 			info.Groups[i] = a.groupsPrefix + group
 		}
 	}
+
+	// check to ensure all required claims are present in the ID token and have matching values.
+	if len(a.requiredClaims) != 0 {
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: nit: if statement is redundant. Ranging over a nil map is already a no-op https://play.golang.org/p/joqoexhvvbN

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 3]
```diff
@@ -323,6 +329,25 @@ func (a *Authenticator) AuthenticateToken(token string) (user.Info, bool, error)
...
+		for claim, value := range a.requiredClaims {
+			if hasRequiredClaim := c.hasClaim(claim); hasRequiredClaim {
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: nit: don't need an `else` block

```
if !c.hasClaim(claim) {
    // return error
}

// compare claim
```

**Review by ericchiang — COMMENTED** [review]
please add another tests with multiple required claims.

File: staging/src/k8s.io/apimachinery/Godeps/Godeps.json (hunk 1 above)
Comment (rithujohn191) [reply staging/src/k8s.io/apimachinery/Godeps/Godeps.json]: reverted all the godep updates. Only ran  `./hack/update-bazel.sh`

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 4]
```diff
@@ -323,6 +329,25 @@ func (a *Authenticator) AuthenticateToken(token string) (user.Info, bool, error)
...
+				var claimValue string
```
[the remark this answers is not in the document]
Comment (rithujohn191) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Added a comment

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go (hunk 3 above)
Comment (rithujohn191) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Done

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go (hunk 2 above)
Comment (rithujohn191) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Done

**Comment by ericchiang** [comment]
cc @kubernetes/sig-auth-pr-reviews

File: pkg/kubeapiserver/options/authentication.go [hunk 5]
```diff
@@ -223,6 +225,10 @@ func (s *BuiltInAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
 			"Comma-separated list of allowed JOSE asymmetric signing algorithms. JWTs with a "+
 			"'alg' header value not in this list will be rejected. "+
 			"Values are defined by RFC 7518 https://tools.ietf.org/html/rfc7518#section-3.1.")
+
+		fs.Var(flag.NewMapStringString(&s.OIDC.RequiredClaims), "oidc-required-claims", ""+
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authentication.go]: will the claim values ever have commas? if they are something like distinguished names, they could. if so, we'd need a different flag type that you repeat rather than split on `,`

Comment (rithujohn191) [reply pkg/kubeapiserver/options/authentication.go]: Glancing through the [standard claims](http://openid.net/specs/openid-connect-core-1_0.html#Claims) it looks like the claim values could possibly have commas (e.g the address claim). I will go ahead and write up a new flag type.

File: staging/src/k8s.io/apiserver/pkg/util/flag/map_string_string.go [hunk 6]
```diff
@@ -23,17 +23,31 @@ import (
...
+// Multiple flag invocations are supported. For example: `--flag "a=foo" --flag "b=bar"`. If this is desired
+// to be the only type invocation `MultiFlagInvocationOnly` should be set to true.
+// Multiple comma-separated key-value pairs in a single invocation are supported if `MultiFlagInvocationOnly`
+// is set to false. For example: `--flag "a=foo,b=bar"`.
 type MapStringString struct {
-	Map         *map[string]string
-	initialized bool
+	Map                     *map[string]string
+	initialized             bool
+	MultiFlagInvocationOnly bool
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/util/flag/map_string_string.go]: `MultiFlagInvocationOnly` ouch. Maybe `NoSplit` instead?

Comment (rithujohn191) [reply staging/src/k8s.io/apiserver/pkg/util/flag/map_string_string.go]: Done

**Comment by ericchiang** [comment]
Thanks for the flag changes. So supplying multiple key values now looks like:

```
--oidc-required-claim=hd=example.com
--oidc-required-claim=signer=CN=Foo,O=Bar
```

/lgtm

File: pkg/kubeapiserver/options/authentication.go [hunk 7]
```diff
@@ -223,6 +225,10 @@ func (s *BuiltInAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
 			"Comma-separated list of allowed JOSE asymmetric signing algorithms. JWTs with a "+
 			"'alg' header value not in this list will be rejected. "+
 			"Values are defined by RFC 7518 https://tools.ietf.org/html/rfc7518#section-3.1.")
+
+		fs.Var(flag.NewMapStringStringNoSplit(&s.OIDC.RequiredClaims), "oidc-required-claim", ""+
+			"A key=value pair that describes a required claim in the ID Token "+
```
[the remark this answers is not in the document]
Comment (rithujohn191) [reply pkg/kubeapiserver/options/authentication.go]: Done.

File: pkg/kubeapiserver/options/authentication.go [hunk 8]
```diff
@@ -223,6 +225,10 @@ func (s *BuiltInAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
...
+			"If set, the claim is verified to be present in the ID Token with matching values.")
```
[the remark this answers is not in the document]
Comment (rithujohn191) [reply pkg/kubeapiserver/options/authentication.go]: Done.

File: pkg/kubeapiserver/options/authentication.go [hunk 9]
```diff
@@ -223,6 +225,10 @@ func (s *BuiltInAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
...
 	}
```
[the remark this answers is not in the document]
Comment (rithujohn191) [reply pkg/kubeapiserver/options/authentication.go]: Done.

**Comment by rithujohn191** [comment]
rebased off master to see if tests will pass
